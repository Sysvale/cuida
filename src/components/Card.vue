<!-- eslint-disable vue/multi-word-component-names -->
<template>
	<CdsBox
		padding="0"
		:clickable="clickable"
		:fluid="fluid"
		@box-click="handleClick"
	>
		<div class="card__extra-container">
			<div class="card__extra">
				<!-- @slot Slot para incluir conteúdo adicional no canto superior direito do card.-->
				<slot name="extra" />
			</div>
		</div>

		<div :class="{'card--horizontal': horizontal}">
			<div
				v-if="hasSlot($slots, 'image')"
			>
				<!-- @slot Slot utilizado para incluir imagens no Card.-->
				<slot name="image" />
			</div>
	
			<div
				v-else-if="imageSrc"
				class="card__image"
			>
				<CdsImage
					:src="imageSrc"
					:width="imageWidthResolver"
					:height="imageHeight"
					:alt="imageAlt"
				/>
			</div>
			
			<div
				v-else-if="hasSlot($slots, 'icon')"
				class="card__icon"
			>
				<!-- @slot Slot utilizado para incluir ícones no Card.-->
				<slot name="icon" />
			</div>

			<div
				class="card__spacer"
			>
				<div
					v-if="hasSlot($slots, 'header')"
					class="card__header"
				>
					<!-- @slot Slot para conteúdo customizado do Header.-->
					<slot name="header" />
				</div>

				<div v-else>
					<p class="card__header">
						{{ title }}
					</p>
				</div>
		
				<div
					v-if="hasSlot($slots, 'body')"
					class="card__body"
				>
					<!-- @slot Slot para conteúdo customizado do Body.-->
					<slot name="body" />
				</div>

				<div v-else>
					<p class="card__body">
						{{ content }}
					</p>
				</div>
		
				<div
					v-if="hasSlot($slots, 'footer')"
					class="card__footer"
				>
					<!-- @slot Slot para conteúdo customizado do Footer.-->
					<slot name="footer" />
				</div>
			</div>
		</div>
	</CdsBox>
</template>

<script setup>
import { computed, defineProps } from 'vue';
import hasSlot from '../utils/methods/hasSlot';
import CdsBox from './Box.vue';
import CdsImage from './Image.vue';

const props = defineProps({
	/**
	* Especifica o título do card. Quando conteúdo é enviado para o slot `Header` o conteúdo dessa prop não é exibido.
	*/
	title: {
		type: String,
		default: '',
	},
	/**
	* Especifica texto do Card. Quando conteúdo é enviado para o slot `Body` o conteúdo dessa prop não é exibido.
	*/
	content: {
		type: String,
		default: '',
	},
	/**
	* Caminho da imagem que vai ser renderizada. Quando conteúdo é enviado para
	o slot `Image` o conteúdo dessa prop não é exibido.
	*/
	imageSrc: {
		type: String,
		default: '',
	},
	/**
	* Descrição em texto da imagem.
	*/
	imageAlt: {
		type: String,
		default: 'imagem do card',
	},
	/**
	* Largura da imagem do card. Tem comportamento sobrescrito quando o card é vertical
	e a prop `fluid` está ativa.
	*/
	imageWidth: {
		type: [String, Number],
		default: 300,
	},
	/**
	* Altura da imagem do card.
	*/
	imageHeight: {
		type: [String, Number],
		default: 180,
	},
	/**
	* Largura do conteúdo do card.
	*/
	bodyWidth: {
		type: [String, Number],
		default: 300,
	},
	/**
	* Torna o alinhamento do Card horizontal.
	*/
	horizontal: {
		type: Boolean,
		default: false,
	},
	/**
	* Ativa ou desativa o clique no componente
	*/
	clickable: {
		type: Boolean,
		default: false,
	},
	/**
	* Ativa ou desativa o comportamento fluido do Card
	*/
	fluid: {
		type: Boolean,
		default: false
	}
})

const emits = defineEmits([
	/**
	 * Evento que indica se o card foi clicado.
	 * @event cardClick
	 * @type {Event}
	 */
	'cardClick'
]);

const cardSpacerMaxWidthResolver = computed(() => {
	if (props.fluid) {
		return '100%';
	}

	return props.horizontal ? 'fit-content' :  `${props.imageWidth}px`;
});

const bodyWidthResolver = computed(() => {
	return `${props.bodyWidth}px`;
})

const imageWidthResolver = computed(() => {
	if (props.fluid && !props.horizontal) return '100%';

	return props.imageWidth
})

function handleClick() {
	if (props.clickable) {
		emits('cardClick', true);
	}
}
</script>

<style lang="scss" scoped>
@use '../assets/sass/tokens/index' as tokens;

.card {
	&__icon {
		padding: tokens.pTRBL(5, 0, 5, 5);
	}
	
	&__extra-container {
		position: relative;
	}
	
	&__extra {
		margin-top: 20px;
		position: absolute;
		right: 16px;
	}
	
	&__image {
		display: flex;
	}
	
	&--horizontal {
		align-items: center;
		display: flex;
	}
	
	&__header {
		@include tokens.subheading-2;
		color: tokens.$n-700;
		margin: tokens.ma(0);
	}
	
	&__body {
		@include tokens.body-2;
		color: tokens.$n-500;
		line-height: 150%;
		margin: tokens.mt(3);
		max-width: v-bind(bodyWidthResolver);
	}
	
	&__spacer {
		padding: tokens.pa(5);
		max-width: v-bind(cardSpacerMaxWidthResolver);
	}
	
	&__footer {
		margin: tokens.mt(0);
	}
}
</style>
