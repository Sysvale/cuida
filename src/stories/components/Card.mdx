import { Meta, Story, Props, Controls, Canvas, Preview, Source } from '@storybook/blocks';
import dedent from 'ts-dedent';
import * as CardStories from './Card.stories';

import Card from '../../components/Card.vue';
import Button from '../../components/Button.vue';
import Badge from '../../components/Badge.vue';
import Icon from '../../components/Icon.vue';

<Meta of={CardStories} />

export const TemplateStandard = (args) => ({
	components: { CdsCard: Card, CdsButton: Button, CdsBadge: Badge },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<cds-card
			v-bind="args"
			@cardClick="logClick"
		>
			<template #header>
				<div style="margin-bottom: 12px">
					Card Title
				</div>
			</template>
			<template #extra>
				<cds-badge>Action</cds-badge>
			</template>
			<template #body>
				Some quick example text to build on the card title and make up the bulk of the card's content.
			</template>
			<template #footer>
				<cds-button size="md" style="margin-top: 16px">
					Action 1
				</cds-button>
			</template>
		</cds-card>
	`,
	methods: {
		logClick(event) {
			console.info('logClick: ', event);
		}
	},
});

export const SimpleCardTemplate = (args) => ({
	components: { CdsCard: Card, CdsButton: Button, CdsBadge: Badge },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<cds-card
			v-bind="args"
		>
		</cds-card>
	`,
});

export const withSlotsTemplate = (args) => ({
	components: { CdsCard: Card, CdsButton: Button, CdsBadge: Badge },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<cds-card
			v-bind="args"
		>
			<template #header>45 <small>gestantes</small></template>
			<template #extra>
				<cds-badge variant="red">Alerta</cds-badge>
			</template>
			<template #body>
				Não realizam consulta há mais de 30 dias
			</template>
			<template #footer>
				<div class="d-flex justify-content-end">
					<small><strong>VER LISTA</strong></small>
				</div>
			</template>
		</cds-card>
	`,
});

export const withIconTemplate = (args) => ({
	components: { CdsCard: Card, CdsButton: Button, CdsIcon: Icon },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<cds-card
			v-bind="args"
		>
			<template #icon>
				<div class="d-flex align-items-center justify-content-center" style="background: #ecf1fe; height: 48px; width: 48px; border-radius: 8px">
					<cds-icon name="home-outline" color="#1A55A8"/>
				</div>
			</template>
			<template #header>
				<div style="font-size: 24px; margin-bottom: 4px"> 32 </div>
			</template>
			<template #body>
				UBSs com dados atualizados
			</template>
		</cds-card>
	`,
});

export const horizontalPropTemplate = (args) => ({
	components: { CdsCard: Card, CdsButton: Button, CdsBadge: Badge },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<cds-card
			v-bind="args"
		>
		</cds-card>
	`,
});

# Card

### Cards são componentes utilizados como containers de propósito geral. À partir das suas props e slots `header`, `body`, `footer`, `image`, `icon` e `extra` é possível criar variações e cards mais específicos, como o DashboardCard.

***

<br />

## Quando usar:

* Para agrupar elementos relacionados entre si.
* Para dar proeminência visual a elementos relacionados entre si.

<br />

## Quando não usar:

* De modo aninhado, com cards dentro de cards.
* Quando o conteúdo encapsulado pelo card ocupar mais da metade da página.

<br />

## Observações:

* ⚠️ Há exemplos adicionais de card abaixo da tabela de props e slots.

## Preview

<Canvas>
  <Story of={CardStories.CardStandard} />
</Canvas>

<Controls />

## **Card simples**

As props `title` `content` e `imageSrc` nos permitem montar um layout simples de Card utilizando apenas props.

<br />

<Story of={CardStories.SimpleCard} />

<br />

<br />

## **Utilizando slots**

Utilizando os slots `title`, `body`, `footer` e `extra` conseguimos criar Cards mais complexos, como o DashboardCard.

<br />

<Story of={CardStories.CardWithSlots} />

<br />

<br />

## **Slot icon**

O Card possui um slot chamado `icon`, que pode ser utilizado para adição de ícones no Card.

⚠️O Slot `image` tem precedência maior que o slot `icon`. Caso os dois sejam utiliados o slot `image` será exibido
e o slot `icon` será ocultado.

<br />

<Story of={CardStories.CardWithIcon} />

<br />

<br />

## **Prop `horizontal`**

Para layouts horizontais de Card, utilize a prop `horizontal`

<br />

<Story of={CardStories.CardWithHorizontalProp} />
