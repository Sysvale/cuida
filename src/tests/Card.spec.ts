// @vitest-environment jsdom
import { describe, test, expect } from 'vitest';
import Card from '../components/Card.vue';
import { mount } from '@vue/test-utils';

describe('Card snapshot test', () => {
	test('renders correctly', () => {
		const wrapper = mount(Card, {
			props: {
				title: 'Card Title',
				content: "Some quick example text to build on the card title and make up the bulk of the card's content.",
			}
		});

		expect(wrapper.html()).toMatchSnapshot();
	});
});
