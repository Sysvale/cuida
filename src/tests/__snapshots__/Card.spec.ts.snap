// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card snapshot test > renders correctly 1`] = `
"<div data-v-f9290560="" data-v-c6c3362a="" class="box--gray">
  <!--
		Evento emitido quando o wrapper é clicado
		@event cds-click
		@type {Event}
	-->
  <div data-v-f9290560="" class="clickable__container--not-clickable box__container">
    <!-- @slot Slot utilizado para renderização do conteúdo interno do Box.-->
    <div data-v-9cc3ee57="" data-v-f9290560="" class="spacer">
      <div data-v-c6c3362a="" class="card__extra-container">
        <div data-v-c6c3362a="" class="card__extra">
          <!-- @slot Slot para incluir conteúdo adicional no canto superior direito do card.-->
        </div>
      </div>
      <div data-v-c6c3362a="" class="">
        <!--v-if-->
        <div data-v-c6c3362a="" class="card__spacer">
          <div data-v-c6c3362a="">
            <p data-v-c6c3362a="" class="card__header">Card Title</p>
          </div>
          <div data-v-c6c3362a="">
            <p data-v-c6c3362a="" class="card__body">Some quick example text to build on the card title and make up the bulk of the card's content.</p>
          </div>
          <!--v-if-->
        </div>
      </div>
    </div>
  </div>
</div>"
`;
